File: experiments/10_edivisive.R
# e-divisive on the raw ordered plot sequences from 10_edivisive.py.
# Reports, per cell: how often a change point is found (sig.lvl 0.05) and
# where (gradient position of the estimated split), against true z* = 0.55.
suppressMessages(library(ecp))
root <- "../data/ediv_sim"
cells <- list(c("THRESH","80"), c("STRENGTH","80"), c("REWIRE","80"),
              c("TRI_NULL","80"), c("REWIRE","160"), c("TRI_NULL","160"))
for (cell in cells) {
  scen <- cell[1]; P <- cell[2]
  found <- 0; locs <- c()
  for (i in 0:9) {
    X <- as.matrix(read.csv(file.path(root, sprintf("%s_%s_%d.csv", scen, P, i))))
    z <- read.csv(file.path(root, sprintf("%s_%s_%d_z.csv", scen, P, i)))[[1]]
    res <- e.divisive(X, sig.lvl = 0.05, R = 199, min.size = 10)
    cps <- res$estimates[c(-1, -length(res$estimates))]  # interior points
    if (length(cps) > 0) { found <- found + 1; locs <- c(locs, z[cps[1]]) }
  }
  cat(sprintf("%9s P=%3s: detected %2d/10", scen, P, found))
  if (length(locs) > 0)
    cat(sprintf("   first-cp gradient position: median %.2f range %.2f-%.2f",
        median(locs), min(locs), max(locs)))
  cat("   (true z* = 0.55)\n")
}

File: experiments/10_edivisive.py
"""Export raw simulated communities for the e-divisive baseline.

E-divisive (Matteson & James 2014) is the statistical incumbent for
multivariate change-point detection: energy-distance tests on the raw
sequence of plot abundance vectors, no summary at all.  A reviewer will ask
why the windowed topological pipeline beats just running that; this
experiment answers with detection rates on the same simulated gradients.
Ten realisations per cell; REWIRE and TRI_NULL also at P=160, the size the
shape summaries need.  Run 10_edivisive.R afterwards.
"""
import importlib
import os

import numpy as np
import pandas as pd

swap = importlib.import_module("08_summary_swap")

OUT = os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__))),
                   "data", "ediv_sim")
os.makedirs(OUT, exist_ok=True)
CELLS = [("THRESH", 80), ("STRENGTH", 80), ("REWIRE", 80), ("TRI_NULL", 80),
         ("REWIRE", 160), ("TRI_NULL", 160)]
for scen, P in CELLS:
    for i in range(10):
        X, z = swap.community(P, scen, np.random.default_rng(600 + i))
        o = np.argsort(z)
        pd.DataFrame(X[:, o].T).to_csv(
            os.path.join(OUT, f"{scen}_{P}_{i}.csv"), index=False)
        pd.Series(z[o]).to_csv(
            os.path.join(OUT, f"{scen}_{P}_{i}_z.csv"), index=False)
print(f"wrote {10 * len(CELLS)} sequences to {OUT}")
